File: ios/Packages/HapiKit/Sources/HapiClient/Chat/PermissionInputs.swift
import Foundation
import HapiProtocol

// Question/option models + input parsers for the two interactive request
// tools, ported from `web/src/components/ToolCard/askUserQuestion.ts`,
// `cursorAskQuestion.ts` and `requestUserInput.ts` via the Android port's
// `PermissionInputs.kt`. Pure functions over `JSONValue` — unit-tested
// alongside the interactor.

public struct AskOption: Equatable, Sendable {
    /// Stable option id (Cursor ACP); falls back to the label on submit.
    public let id: String?
    public let label: String
    public let description: String?

    public init(id: String?, label: String, description: String?) {
        self.id = id
        self.label = label
        self.description = description
    }
}

public struct AskQuestion: Equatable, Sendable {
    /// Stable question id (Cursor ACP); falls back to the index on submit.
    public let id: String?
    public let header: String?
    public let question: String
    public let options: [AskOption]
    public let multiSelect: Bool

    public init(id: String?, header: String?, question: String, options: [AskOption], multiSelect: Bool) {
        self.id = id
        self.header = header
        self.question = question
        self.options = options
        self.multiSelect = multiSelect
    }

    /// The flat-answers key: stable id when present, else the index (web parity).
    public func answerKey(index: Int, useStableIds: Bool) -> String {
        if useStableIds, let id, !id.isEmpty {
            return id
        }
        return String(index)
    }
}

public func isCursorAskQuestionToolName(_ toolName: String) -> Bool {
    toolName == "CursorAskQuestion"
}

private func trimmedString(_ value: JSONValue?) -> String? {
    value?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines)
}

/// `parseAskUserQuestionInput` / `parseCursorAskQuestionInput` merged: the
/// Cursor dialect adds `prompt`/`title`/`allowMultiple` synonyms and stable
/// ids; both collapse to the same `AskQuestion` list.
public func parseAskUserQuestions(_ input: JSONValue?, cursorDialect: Bool) -> [AskQuestion] {
    guard let root = input?.objectValue,
          let rawQuestions = root["questions"]?.arrayValue else {
        return []
    }
    let requestTitle = cursorDialect ? (trimmedString(root["title"]) ?? "") : ""

    var questions: [AskQuestion] = []
    for raw in rawQuestions {
        guard let object = raw.objectValue else { continue }

        let question: String
        let header: String
        let multiSelect: Bool
        let questionId: String?
        if cursorDialect {
            question = trimmedString(object["prompt"]) ?? trimmedString(object["question"]) ?? ""
            header = trimmedString(object["title"]) ?? trimmedString(object["header"]) ?? ""
            multiSelect = object["allowMultiple"]?.boolValue == true
                || object["multiSelect"]?.boolValue == true
            let rawId = trimmedString(object["id"]) ?? ""
            questionId = rawId.isEmpty ? String(questions.count) : rawId
        } else {
            question = trimmedString(object["question"]) ?? ""
            header = trimmedString(object["header"]) ?? ""
            multiSelect = object["multiSelect"]?.boolValue == true
            questionId = nil
        }

        var options: [AskOption] = []
        for rawOption in object["options"]?.arrayValue ?? [] {
            guard let optionObject = rawOption.objectValue else { continue }
            let label: String
            if cursorDialect {
                label = trimmedString(optionObject["label"]) ?? trimmedString(optionObject["id"]) ?? ""
            } else {
                label = trimmedString(optionObject["label"]) ?? ""
            }
            guard !label.isEmpty else { continue }
            let optionId: String?
            if cursorDialect {
                let rawId = trimmedString(optionObject["id"]) ?? ""
                optionId = rawId.isEmpty ? label : rawId
            } else {
                optionId = nil
            }
            let description = cursorDialect
                ? nil
                : trimmedString(optionObject["description"]).flatMap { $0.isEmpty ? nil : $0 }
            options.append(AskOption(id: optionId, label: label, description: description))
        }

        if question.isEmpty && options.isEmpty { continue }

        let effectiveHeader = header.isEmpty ? (requestTitle.isEmpty ? nil : requestTitle) : header
        questions.append(AskQuestion(
            id: questionId,
            header: effectiveHeader,
            question: question,
            options: options,
            multiSelect: multiSelect
        ))
    }
    return questions
}

public struct RequestUserInputQuestion: Equatable, Sendable {
    public let id: String
    public let question: String
    public let required: Bool
    public let multiple: Bool
    public let options: [AskOption]
    public let placeholder: String?
    public let prefill: String?

    public init(
        id: String,
        question: String,
        required: Bool,
        multiple: Bool,
        options: [AskOption],
        placeholder: String?,
        prefill: String?
    ) {
        self.id = id
        self.question = question
        self.required = required
        self.multiple = multiple
        self.options = options
        self.placeholder = placeholder
        self.prefill = prefill
    }
}

/// `parseRequestUserInputInput` (the URL-confirmation flow is web-only).
public func parseRequestUserInputQuestions(_ input: JSONValue?) -> [RequestUserInputQuestion] {
    guard let root = input?.objectValue,
          let rawQuestions = root["questions"]?.arrayValue else {
        return []
    }

    var questions: [RequestUserInputQuestion] = []
    for raw in rawQuestions {
        guard let object = raw.objectValue else { continue }
        let id = trimmedString(object["id"]) ?? ""
        guard !id.isEmpty else { continue }

        var options: [AskOption] = []
        for rawOption in object["options"]?.arrayValue ?? [] {
            guard let optionObject = rawOption.objectValue else { continue }
            let label = trimmedString(optionObject["label"]) ?? ""
            guard !label.isEmpty else { continue }
            options.append(AskOption(
                id: nil,
                label: label,
                description: trimmedString(optionObject["description"]).flatMap { $0.isEmpty ? nil : $0 }
            ))
        }

        questions.append(RequestUserInputQuestion(
            id: id,
            question: trimmedString(object["question"]) ?? "",
            required: object["required"]?.boolValue != false,
            multiple: object["multiple"]?.boolValue == true,
            options: options,
            placeholder: object["placeholder"]?.stringValue,
            prefill: object["prefill"]?.stringValue
        ))
    }
    return questions
}

/// `formatRequestUserInputAnswers` value building for ONE field: selected
/// option labels plus a trailing `user_note: <text>` entry when a note was
/// typed. The nested `{answers: [...]}` wrapper is applied by the interactor.
public func requestUserInputAnswerValues(selected: [String], note: String) -> [String] {
    var values = selected
    let trimmed = note.trimmingCharacters(in: .whitespacesAndNewlines)
    if !trimmed.isEmpty {
        values.append("user_note: \(trimmed)")
    }
    return values
}

/// `isRequestUserInputQuestionAnswered`.
public func isRequestUserInputAnswered(
    _ question: RequestUserInputQuestion,
    selected: [String],
    note: String
) -> Bool {
    if !question.required { return true }
    if !question.options.isEmpty { return !selected.isEmpty }
    return !note.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
}

File: android/app/src/main/kotlin/app/hapi/companion/feature/chat/permissions/PermissionInputs.kt
package app.hapi.companion.feature.chat.permissions

import app.hapi.protocol.wire.arrayOrNull
import app.hapi.protocol.wire.boolOrNull
import app.hapi.protocol.wire.objOrNull
import app.hapi.protocol.wire.stringOrNull
import kotlinx.serialization.json.JsonElement

/**
 * Question/option models + input parsers for the two interactive request
 * tools, ported from `web/src/components/ToolCard/askUserQuestion.ts`,
 * `cursorAskQuestion.ts` and `requestUserInput.ts`. Pure JVM — unit-tested
 * alongside the ViewModel.
 */

data class AskOption(
    /** Stable option id (Cursor ACP); falls back to the label on submit. */
    val id: String?,
    val label: String,
    val description: String?,
)

data class AskQuestion(
    /** Stable question id (Cursor ACP); falls back to the index on submit. */
    val id: String?,
    val header: String?,
    val question: String,
    val options: List<AskOption>,
    val multiSelect: Boolean,
) {
    /** The flat-answers key: stable id when present, else the index (web parity). */
    fun answerKey(index: Int, useStableIds: Boolean): String =
        if (useStableIds && !id.isNullOrBlank()) id else index.toString()
}

fun isCursorAskQuestionToolName(toolName: String): Boolean = toolName == "CursorAskQuestion"

/**
 * `parseAskUserQuestionInput` / `parseCursorAskQuestionInput` merged: the
 * Cursor dialect adds `prompt`/`title`/`allowMultiple` synonyms and stable
 * ids; both collapse to the same [AskQuestion] list.
 */
fun parseAskUserQuestions(input: JsonElement?, cursorDialect: Boolean): List<AskQuestion> {
    val root = input.objOrNull ?: return emptyList()
    val rawQuestions = root["questions"].arrayOrNull ?: return emptyList()
    val requestTitle = if (cursorDialect) root["title"].stringOrNull?.trim().orEmpty() else ""

    val questions = mutableListOf<AskQuestion>()
    for (raw in rawQuestions) {
        val obj = raw.objOrNull ?: continue

        val question = if (cursorDialect) {
            obj["prompt"].stringOrNull?.trim() ?: obj["question"].stringOrNull?.trim().orEmpty()
        } else {
            obj["question"].stringOrNull?.trim().orEmpty()
        }
        val header = if (cursorDialect) {
            obj["title"].stringOrNull?.trim() ?: obj["header"].stringOrNull?.trim().orEmpty()
        } else {
            obj["header"].stringOrNull?.trim().orEmpty()
        }
        val multiSelect = if (cursorDialect) {
            obj["allowMultiple"].boolOrNull == true || obj["multiSelect"].boolOrNull == true
        } else {
            obj["multiSelect"].boolOrNull == true
        }
        val questionId = if (cursorDialect) {
            obj["id"].stringOrNull?.trim()?.takeIf { it.isNotEmpty() } ?: questions.size.toString()
        } else {
            null
        }

        val options = mutableListOf<AskOption>()
        for (rawOption in obj["options"].arrayOrNull ?: emptyList()) {
            val optionObj = rawOption.objOrNull ?: continue
            val label = if (cursorDialect) {
                optionObj["label"].stringOrNull?.trim() ?: optionObj["id"].stringOrNull?.trim().orEmpty()
            } else {
                optionObj["label"].stringOrNull?.trim().orEmpty()
            }
            if (label.isEmpty()) continue
            val optionId = if (cursorDialect) {
                optionObj["id"].stringOrNull?.trim()?.takeIf { it.isNotEmpty() } ?: label
            } else {
                null
            }
            val description = if (cursorDialect) null else optionObj["description"].stringOrNull?.trim()
            options += AskOption(id = optionId, label = label, description = description?.takeIf { it.isNotEmpty() })
        }

        if (question.isEmpty() && options.isEmpty()) continue

        questions += AskQuestion(
            id = questionId,
            header = header.ifEmpty { requestTitle.ifEmpty { null } },
            question = question,
            options = options,
            multiSelect = multiSelect,
        )
    }
    return questions
}

data class RequestUserInputQuestion(
    val id: String,
    val question: String,
    val required: Boolean,
    val multiple: Boolean,
    val options: List<AskOption>,
    val placeholder: String?,
    val prefill: String?,
)

/** `parseRequestUserInputInput` (the URL-confirmation flow is web-only). */
fun parseRequestUserInputQuestions(input: JsonElement?): List<RequestUserInputQuestion> {
    val root = input.objOrNull ?: return emptyList()
    val rawQuestions = root["questions"].arrayOrNull ?: return emptyList()

    val questions = mutableListOf<RequestUserInputQuestion>()
    for (raw in rawQuestions) {
        val obj = raw.objOrNull ?: continue
        val id = obj["id"].stringOrNull?.trim().orEmpty()
        if (id.isEmpty()) continue

        val options = mutableListOf<AskOption>()
        for (rawOption in obj["options"].arrayOrNull ?: emptyList()) {
            val optionObj = rawOption.objOrNull ?: continue
            val label = optionObj["label"].stringOrNull?.trim().orEmpty()
            if (label.isEmpty()) continue
            options += AskOption(
                id = null,
                label = label,
                description = optionObj["description"].stringOrNull?.trim()?.takeIf { it.isNotEmpty() },
            )
        }

        questions += RequestUserInputQuestion(
            id = id,
            question = obj["question"].stringOrNull?.trim().orEmpty(),
            required = obj["required"].boolOrNull != false,
            multiple = obj["multiple"].boolOrNull == true,
            options = options,
            placeholder = obj["placeholder"].stringOrNull,
            prefill = obj["prefill"].stringOrNull,
        )
    }
    return questions
}

/**
 * `formatRequestUserInputAnswers` value building for ONE field: selected
 * option labels plus a trailing `user_note: <text>` entry when a note was
 * typed. The nested `{answers: [...]}` wrapper is applied by the ViewModel.
 */
fun requestUserInputAnswerValues(selected: List<String>, note: String): List<String> {
    val values = selected.toMutableList()
    val trimmed = note.trim()
    if (trimmed.isNotEmpty()) values += "user_note: $trimmed"
    return values
}

/** `isRequestUserInputQuestionAnswered`. */
fun isRequestUserInputAnswered(
    question: RequestUserInputQuestion,
    selected: List<String>,
    note: String,
): Boolean {
    if (!question.required) return true
    if (question.options.isNotEmpty()) return selected.isNotEmpty()
    return note.trim().isNotEmpty()
}
